### PR TITLE
Split up attribute parsing code and move data types to `rustc_attr_data_structures`

### PR DESCRIPTION
**Pull request by jdonszelmann**
This change renames `rustc_attr` to `rustc_attr_parsing`, and splits up the parsing code. At the same time, all the data types used move to `rustc_attr_data_structures`. This is in preparation of also having a third crate: `rustc_attr_validation`

I initially envisioned this as two separate PRs, but I think doing it in one go reduces the number of ways others would have to rebase their changes on this. However, I can still split them.

r? @oli-obk (we already discussed how this is a first step in a larger plan)

For a more detailed plan on how attributes are going to change, see https://github.com/rust-lang/rust/issues/131229

Edit: this looks like a giant PR, but the changes are actually rather trivial. Each commit is reviewable on its own, and mostly moves code around. No new logic is added.

### CODE REVIEW

**Comment by jdonszelmann** [comment]
@rustbot label +A-attributes

For the rustdoc changes, I only made that code depend on `rustc_attr_parsing` instead of `rustc_attr`. No actual logic changed.

**Comment by jdonszelmann** [comment]
well, my last rebase had no conflicts, but clearly it didn't work very well anyway...

File: a.out
Comment (oli-obk) [inline a.out:1]: accidentally committed a test file

Comment (oli-obk) [reply a.out:1]: (even though it was removed in the second commit)

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by oli-obk** [comment]
I reviewed some changes and the overall change. I'm just gonna assume all the individual parts have been moved correctly (as no tests have been touched, and the downstream usages are all just import changes, this appears so)

@bors delegate+

r=me with the nit for a clean history

File: a.out
Comment (jdonszelmann) [reply a.out:1]: oops, yea I noticed but should've removed it from the first one haha

**Comment by jieyouxu** [comment]
(conflict-check)

**Comment by jdonszelmann** [comment]
(I'll see if the next few can be less conflict prone lol)

**Pull request merged**